### ISSUE TITLE
How to describe unification done in match constructs?

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#12445]
Is the following, for the doc, an accurate description of how `match` does unification?  I added the "non-linear unification", not sure that's right.  Also would be nice to get an example of second-order unification.

   Matching is done with non-linear unification: if a
   metavariable occurs more than once, each occurrence must match the same
   expression.  Matching is first-order except on variables of the form :n:`@?@ident`
   that occur in the head position of an application. For these variables,
   matching is second-order and returns a functional term.

@ppedrot
cc: @JasonGross

**Comment by SkySkimmer** [comment]
I guess you mean ltac match?

**Comment by JasonGross** [comment]
Yes, this is about Ltac matching.  My understanding is that non-linear matching is done up to syntactic equality.  I'm not sure what happens in the following cases:
- the same evar with two substitutions which are not syntacticly equal but do unify
- the same inductive with different universes when there's no inconsistency in equating the universes

Furthermore, I believe @ppedrot may have had some (other?) examples where noon-linear occurrences are not just syntacticly matched?

As for second-order unification, matching `forall x : Nat, x = 1` against the pattern `forall x, @?P x` will succeed and instantiate `P` with `fun x : nat => x = 1`.  This is second-order because `x = 1` is not an application of a function symbol to a list of arguments ending with `x`.

**Comment by ppedrot** [comment]
@JasonGross It's more complex than that:
- inside the same term (e.g. if you have `match goal with [ |- ?P -> ?P ] => ... end`) non-linear occurrences are matched up to syntactic equality
- across terms (e.g. if you have `match goal with [ H : ?P |- ?P ]` => ... end) non-linear occurrences are matched up to conversion
